### PR TITLE
Turn on xfail_strict=true

### LINKED ISSUE #7386: Remove xfail condition from now passing tests
Several tests in the cuDF Python codebase were marked `xfail` during initial development but now pass. We should remove the `xfail` condition from these tests.

### PR DESCRIPTION
**Pull request by wence-**
## Description

Systematically work through all currently XPASSing tests and remove
them. Mostly being more careful with the set of parameters that are
marked as XFAIL. In many cases the bug was already fixed!

Furthermore, since programmatic pytest.xfail doesn't really do
anything useful (subsequent code is not run, so we never notice if
something is fixed), add advice in the developer docs, and expunge all
use of programmatic pytest.xfail (enforced via pre-commit). This leads
to slightly worse code in some places in the test suite, but on
balance I think it is worth it since we will now complain loudly when
a previously xfailing feature is fixed.

Close #7386.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/core/reshape.py
Comment (wence-) [inline python/cudf/cudf/core/reshape.py:280]: This fixes #12132, I'll spin it out.

Comment (wence-) [reply python/cudf/cudf/core/reshape.py:280]: https://github.com/rapidsai/cudf/pull/12247

File: python/cudf/cudf/core/tools/datetimes.py
Comment (wence-) [inline python/cudf/cudf/core/tools/datetimes.py:903]: This comment still applies, so I should reinstate it.

File: python/cudf/cudf/core/tools/datetimes.py
Comment (wence-) [inline python/cudf/cudf/core/tools/datetimes.py:978]: I thought these would fix #12133, but they did not, so I should spin them out I think.

File: python/cudf/cudf/testing/_utils.py
Comment (wence-) [inline python/cudf/cudf/testing/_utils.py:343]: Never used anywhere.

File: python/cudf/cudf/tests/test_array_ufunc.py
Comment (wence-) [inline python/cudf/cudf/tests/test_array_ufunc.py:82]: This is a pattern I have to introduce in a bunch of places, which I don't _really_ like, but it's better than xfail which is kind of "wrong"?

File: python/cudf/cudf/tests/test_array_ufunc.py
Comment (wence-) [inline python/cudf/cudf/tests/test_array_ufunc.py:79]: Always true, since we caught attibuteerror, so
```suggestion
```

File: python/cudf/cudf/tests/test_array_ufunc.py
Comment (wence-) [inline python/cudf/cudf/tests/test_array_ufunc.py:90]: It was, but we didn't notice!

File: python/cudf/cudf/tests/test_array_ufunc.py
Comment (wence-) [inline python/cudf/cudf/tests/test_array_ufunc.py:95]: We're expecting an `AssertionError` in pandas < 1.5 rather than `xfailing` programmatically; this pattern repeats...

File: python/cudf/cudf/tests/test_array_ufunc.py
Comment (wence-) [inline python/cudf/cudf/tests/test_array_ufunc.py:288]: Just an indentation change due to the contextmanager.

File: python/cudf/cudf/tests/test_array_ufunc.py
Comment (wence-) [inline python/cudf/cudf/tests/test_array_ufunc.py:367]:
```suggestion
```

File: python/cudf/cudf/tests/test_array_ufunc.py
Comment (wence-) [inline python/cudf/cudf/tests/test_array_ufunc.py:375]: It turns out _some_ of the functions work in pandas when indexed...

File: python/cudf/cudf/tests/test_binops.py
Comment (wence-) [inline python/cudf/cudf/tests/test_binops.py:913]: Spinning the bool case out into a separate test.

File: python/cudf/cudf/tests/test_udf_masked_ops.py
Comment (wence-) [inline python/cudf/cudf/tests/test_udf_masked_ops.py:492]: Pandas is more correct if one is to treat NA like nan.

File: python/cudf/cudf/tests/test_udf_masked_ops.py
Comment (wence-) [inline python/cudf/cudf/tests/test_udf_masked_ops.py:508]: and flake8 complains if you write `constant is 1`

File: python/dask_cudf/dask_cudf/tests/test_groupby.py
Comment (wence-) [inline python/dask_cudf/dask_cudf/tests/test_groupby.py:103]: This was only partially fixed, but enough for this test.

File: python/cudf/cudf/tests/test_udf_masked_ops.py
Comment (brandon-b-miller) [reply python/cudf/cudf/tests/test_udf_masked_ops.py:492]: Many long discussions of this in the past. Short answer is that it's big problem which is also hard to solve because libcudf will never special case single values like this. So we end up having to introspect data in python which has historically been a non-starter. Won't hijack this PR but happy to discuss further.

File: python/cudf/cudf/tests/test_array_ufunc.py
Comment (vyasr) [reply python/cudf/cudf/tests/test_array_ufunc.py:82]: Would it help to introduce something like the `expect_warning_if` contextmanager that I just created, except in this case for exception handling? https://github.com/rapidsai/cudf/blob/branch-23.02/python/cudf/cudf/testing/_utils.py#L393

Comment (vyasr) [reply python/cudf/cudf/tests/test_array_ufunc.py:82]: That may also help reduce or usage of the `nullcontext`.

File: python/cudf/cudf/tests/test_udf_masked_ops.py
Comment (wence-) [reply python/cudf/cudf/tests/test_udf_masked_ops.py:492]: I note there _is_ precedent for post-hoc inspection (with integer division by zero). Arguably checking for NA is easier because we can just inspect `null_count` which doesn't require a scan of the column every time.

File: python/dask_cudf/dask_cudf/tests/test_groupby.py
Comment (wence-) [reply python/dask_cudf/dask_cudf/tests/test_groupby.py:103]: Full fix (dask/dask#9716) is now in.

File: docs/cudf/source/developer_guide/testing.md
Comment (vyasr) [inline docs/cudf/source/developer_guide/testing.md:196]: One note here, I believe that it should be possible to emulate the expected behavior of `pytest.xfail` by doing something like
```
@pytest.mark.parametrize(x, [1, 2, 3])
def test_x(x):
    if x > 1:
        requests.applymarker(pytest.mark.xfail)
```
I wouldn't encourage this usage, but maybe there are some cases where something like that is needed instead of the strictly static mark approach.

File: python/cudf/cudf/core/tools/datetimes.py
Comment (wence-) [reply python/cudf/cudf/core/tools/datetimes.py]: Done.

File: python/cudf/cudf/tests/test_array_ufunc.py
Comment (wence-) [reply python/cudf/cudf/tests/test_array_ufunc.py:89]: In the end, I went with the suggestion below of using `request.applymarker` since that feels a bit more ergonomic.

File: python/cudf/cudf/tests/test_array_ufunc.py
Comment (wence-) [reply python/cudf/cudf/tests/test_array_ufunc.py]: switched to `requests.applymarker(pytest.mark.xfail(condition=when, reason="foo"))`.

File: python/cudf/cudf/tests/test_array_ufunc.py
Comment (wence-) [reply python/cudf/cudf/tests/test_array_ufunc.py]: And gone again.

File: python/cudf/cudf/tests/test_array_ufunc.py
Comment (vyasr) [reply python/cudf/cudf/tests/test_array_ufunc.py:89]: Sounds good. That works much better for failures than warnings.

File: python/cudf/cudf/tests/pytest.ini
Comment (bdice) [inline python/cudf/cudf/tests/pytest.ini:6]: Can we do the same for other packages in the repo? I would expect that dask-cudf / custreamz / cudf-kafka / strings_udf won't have as many problems as cudf. This is fine to do in a follow-up PR, I just want testing conventions to be consistent in the repo.

File: python/cudf/cudf/tests/test_csv.py
Comment (bdice) [reply python/cudf/cudf/tests/test_csv.py:1371]: This is a good question. CI currently tests with the latest pandas version that satisfies our dependency pinning. These version-conditional statements are all throughout our tests but I strongly doubt that any non-latest version would actually pass our tests, since we never verify that. I would recommend we discuss this in our next Python meeting. @quasiben ran into this before and we had a conversation about it here: https://github.com/rapidsai/cudf/pull/12150

File: python/cudf/cudf/tests/test_groupby.py
Comment (bdice) [inline python/cudf/cudf/tests/test_groupby.py:1066]:
```suggestion
                "Pandas NaN Rounding will fail nearest interpolation at 0.5"
```

**Review by bdice — APPROVED** [review]
I have a few small comments, otherwise LGTM.

File: python/cudf/cudf/tests/test_array_ufunc.py
Comment (vyasr) [inline python/cudf/cudf/tests/test_array_ufunc.py:251]: I think the need to applymarker this frequently is an indication that these tests are probably overparametrized. I'm guilty of this since I wrote most of the problematic tests in this file, but it's also a problem that's pretty prevalent throughout the package. Once this PR is merged we may want to follow up with splitting these up.

File: python/cudf/cudf/tests/test_binops.py
Comment (vyasr) [inline python/cudf/cudf/tests/test_binops.py:954]: Rather than marking each parameter, why not just mark the test so that you can parametrize with [False, True]?

File: python/cudf/cudf/tests/test_concat.py
Comment (vyasr) [inline python/cudf/cudf/tests/test_concat.py:1035]: I guess this has just been xpassing for a while?

File: python/cudf/cudf/tests/test_csv.py
Comment (vyasr) [reply python/cudf/cudf/tests/test_csv.py:1371]: Ah this looks like almost exactly the same comment that I had in my review. I agree, let's discuss it in the next meeting and see if we can remove these.

File: python/cudf/cudf/tests/test_dataframe.py
Comment (vyasr) [reply python/cudf/cudf/tests/test_dataframe.py:290]: Do you know what changed in pandas for them to pass?

File: python/cudf/cudf/tests/test_orc.py
Comment (vyasr) [inline python/cudf/cudf/tests/test_orc.py:1764]: Can we apply mark.xfail to the whole test to simplify the parametrization to not need `pytest.param`?

File: python/cudf/cudf/tests/test_parquet.py
Comment (vyasr) [reply python/cudf/cudf/tests/test_parquet.py:2187]: Should we delete this to avoid a potential xfail? I assume there's a bug tracking this issue?

File: python/cudf/cudf/tests/test_parquet.py
Comment (vyasr) [inline python/cudf/cudf/tests/test_parquet.py:2244]: One more case of mark the test to avoid `pytest.param`?

File: python/cudf/cudf/tests/test_timedelta.py
Comment (vyasr) [inline python/cudf/cudf/tests/test_timedelta.py:906]: Let's not use xfail since it is easy to conflate pytest xfails
```suggestion
        # Expect exceptions for certain combinations
```

File: python/cudf/cudf/tests/test_timedelta.py
Comment (vyasr) [inline python/cudf/cudf/tests/test_timedelta.py:1000]:
```suggestion
        # Expect exceptions for certain combinations
```

File: python/cudf/cudf/tests/test_udf_masked_ops.py
Comment (vyasr) [reply python/cudf/cudf/tests/test_udf_masked_ops.py:492]: Probably best to follow up on this issue separately from this PR.

**Review by vyasr — CHANGES_REQUESTED** [review]
This is great to see! Just a few more things to do I think.

File: python/cudf/cudf/tests/test_array_ufunc.py
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_array_ufunc.py:62]: Historically we have been doing this version-based xfailing when we do an upgrade in cudf, but the question here seems to be how do we phase out the very very old versions we don't test or match against anymore. I'm in favor of stripping away very old version-based xfails, but we should probably discuss the pros and cons of doing so with internal teams too.
My pros:
1. Quality of life improvement and lesser ambiguity in tests.
2. Easy to review

Cons:
1. Someone who tests latest cudf with an old version of pandas(say 1.0) will not be able to get all xfails passing. But the tests that are not being xfailed too will anyway fail because our latest cudf is actually trying to match pandas-1.5 rather than pandas-1.0, so this isn't a huge issue. 

Added this to our weekly agenda to discuss what the team feels about on this issue. @quasiben ran into this recently too.

File: python/cudf/cudf/tests/pytest.ini
Comment (wence-) [reply python/cudf/cudf/tests/pytest.ini:6]: Opened an issue.

File: python/cudf/cudf/tests/test_binops.py
Comment (wence-) [reply python/cudf/cudf/tests/test_binops.py:954]: Good point; done.

File: python/cudf/cudf/tests/test_array_ufunc.py
Comment (wence-) [reply python/cudf/cudf/tests/test_array_ufunc.py:251]: Opened an issue.

File: python/cudf/cudf/tests/test_concat.py
Comment (wence-) [reply python/cudf/cudf/tests/test_concat.py:1035]: Yep :)

File: python/cudf/cudf/tests/test_orc.py
Comment (wence-) [reply python/cudf/cudf/tests/test_orc.py:1764]: Only some of these parameters fail, but I should split out the ones that don't...

File: python/cudf/cudf/tests/test_parquet.py
Comment (wence-) [reply python/cudf/cudf/tests/test_parquet.py:2187]: Yes: https://github.com/rapidsai/cudf/issues/12243; I think it is ok (if we fix the first bug, we'll get an immediate fail here which will prompt us to fix things)

File: python/cudf/cudf/tests/test_parquet.py
Comment (wence-) [reply python/cudf/cudf/tests/test_parquet.py:2244]: Note that the `False` case works fine, so I need to split things out... (done).

File: python/cudf/cudf/tests/test_timedelta.py
Comment (wence-) [reply python/cudf/cudf/tests/test_timedelta.py:906]: Fixed by going to the applymarker approach

File: python/cudf/cudf/tests/test_timedelta.py
Comment (wence-) [reply python/cudf/cudf/tests/test_timedelta.py:1000]: Switched to applymarker approach

File: python/cudf/cudf/tests/test_binops.py
Comment (brandon-b-miller) [reply python/cudf/cudf/tests/test_binops.py:1751]: I think that issue might be superseded by https://github.com/pandas-dev/pandas/issues/46877

File: python/cudf/cudf/tests/test_dataframe.py
Comment (wence-) [reply python/cudf/cudf/tests/test_dataframe.py:290]: No clue sorry, this was "mindless" condition based on what I observe.

File: python/cudf/cudf/tests/test_timedelta.py
Comment (wence-) [inline python/cudf/cudf/tests/test_timedelta.py:931]: I thought all the floor division issues were fixed, but it turns out there was special-casing in timedelta.__floordiv__; I'll handle the _fix_ separately.

File: python/cudf/cudf/tests/test_udf_masked_ops.py
Comment (wence-) [inline python/cudf/cudf/tests/test_udf_masked_ops.py:815]: All of the test of this file is whitespace change, I grouped the string udf tests into a class to make skipping them easier.

File: python/dask_cudf/dask_cudf/tests/test_groupby.py
Comment (wence-) [inline python/dask_cudf/dask_cudf/tests/test_groupby.py:109]: I'll go round again and do this properly as part of fixing up the other packages.

**Comment by wence-** [comment]
> This is great to see! Just a few more things to do I think.

Thanks, I think I hit them all

File: python/cudf/cudf/tests/test_orc.py
Comment (vyasr) [reply python/cudf/cudf/tests/test_orc.py:1745]: Oh I think the GH GUI just wasn't quite readable enough for me to notice the ones that didn't fail or I wouldn't have asked. Thanks for splitting though!

File: python/cudf/cudf/tests/test_parquet.py
Comment (vyasr) [reply python/cudf/cudf/tests/test_parquet.py:2230]: Thanks again... the nested parens really make this easy to miss.

**Review by vyasr — APPROVED** [review]
Great!

**Comment by vyasr** [comment]
@wence- I took care of merging the latest 23.02, which should unblock tests here.

**Comment by wence-** [comment]
Final test fails were real, we fixed a bug in dask that was previously an xfail in dask-cudf.

File: python/cudf/cudf/tests/test_array_ufunc.py
Comment (brandon-b-miller) [reply python/cudf/cudf/tests/test_array_ufunc.py:62]: Are our tests considered user facing? IIRC this was invented to get us through some kind of complex CI situation where different test containers ended up pulling different versions of pandas that were allowed within the window. To me, support for a certain range of pandas versions means "your cuDF code will work and produce the same answer, no matter what version of pandas you use within this window" and not much about our unit testing. 

I think viewed from that lens, what this is pointing to is that even though it's very convenient to use pandas to generate "expected" values in most of our tests, it's not actually right to do that because we're aiming for a moving target. Until we stop doing that I suspect we might end up needing flags like this from time to time. I'd rather see us long term take those tests and really take some time to define together with pandas exactly what that API should do, and then write pandas out of the test completely.

Comment (vyasr) [reply python/cudf/cudf/tests/test_array_ufunc.py:62]: Agreed. #6435 is relevant here too.

**Comment by vyasr** [comment]
@gpucibot merge

**Review by quasiben — APPROVED** [review]

**Pull request merged**
